### PR TITLE
File based fixtures

### PR DESCRIPTION
**Pull request by gnott**
There were some existing tests for the ``xmlio.py`` module which loaded file fixtures to run tests for subject tag rewriting. They were only reading XML content, and I changed it anyway in the end to not use ``StringIO`` anymore, which seemed to be the most compatible for supporting Python 2 and 3.

The primary addition here is in ``read_fixture()`` and how it can read ``OrderedDict()`` data from python fixtures. After looking at and considering some other options (``eval``, ``pickle``, etc.) I arrived at this prototype for review. It uses ``importlib``, importing the fixture as a module (if the file name ends in ``.py``) and then returning the ``expected`` property. The intent was to avoid loading JSON data so we can be sure the order of each value remains the same when comparing what is returned by the parser to the test fixture.

In the ``tests/fixtures/`` folder, each test case has its own folder, and in there a suggested convention is to name the files with ``_01``, ``_02``, etc. in the file name, just to group the input and expected files together in a quick way, rather than to name each pair of files based on what edge case they represent. The numeric file naming is not a requirement, but it can probably work as a quick way to convert existing tests cases to use file fixtures.

I welcome comments on this first demonstration, with the aim that what we approve here I will continue to move all the ugly tests to use file fixtures over time.

### CODE REVIEW

File: elifetools/tests/fixtures/test_authors_json/content_02.xml [hunk 1]
```diff
@@ -0,0 +1 @@
+<root xmlns:xlink="http://www.w3.org/1999/xlink"><front><article-meta><contrib-group><contrib contrib-type="author"><collab>ICGC Breast Cancer Group</collab></contrib></contrib-group></article-meta></front></root>
```
Comment (giorgiosironi) [inline elifetools/tests/fixtures/test_authors_json/content_02.xml]: I'm not sure if it changes anything about the result, but if possible I would try to store formatted XML content so that it's easy to view in an editor?

File: elifetools/tests/fixtures/test_authors_json/content_01_expected.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+from collections import OrderedDict
```
Comment (giorgiosironi) [inline elifetools/tests/fixtures/test_authors_json/content_01_expected.py]: This file seems minimal, which is important given there are going to be many of this

File: elifetools/file_utils/__init__.py [hunk 3]
```diff
@@ -19,10 +20,22 @@ def sample_xml(filename):
...
+
 def read_fixture(folder_name, filename):
-    full_filename = os.path.join(fixture_folder(folder_name), filename)
-    with open(full_filename) as file_fp:
-        return unicode_value(file_fp.read())
+    full_filename = fixture_file(folder_name, filename)
+    if full_filename.endswith('.py'):
+        # import the fixture and return the value of expected
+        module_name = fixture_module_name(folder_name, filename)
+        mod = importlib.import_module(module_name)
```
Comment (giorgiosironi) [inline elifetools/file_utils/__init__.py]: could check that `mod.expected` exists (`getattr`?) to make it foolproof

File: elifetools/tests/fixtures/test_authors_json/content_02.xml (hunk 1 above)
Comment (gnott) [reply elifetools/tests/fixtures/test_authors_json/content_02.xml]: Almost all of the XML we parse is collapsed on one line, although that doesn't mean all the specific test examples must also be that way. 

The places where extra new lines, spaces or tabs can make a difference is where inline tags in a paragraph are wrapped to the following line (using some XML pretty output) and as a result we will get extra whitespace in the results. For one example is to look at the eLife 00666 kitchen sink, and it has large strings of spaces where sentences span multiple lines in the XML file.

All that said, I'm happy to manually prettify the XML fixtures a certain amount, as long as it doesn't affect the output by adding extra spaces.

File: elifetools/file_utils/__init__.py (hunk 3 above)
Comment (gnott) [reply elifetools/file_utils/__init__.py]: Good idea, I'm committing with an assertion the property exists in the module before continuing.
